fixes date filtering

eshiol opened the pull request:
### Summary of changes
Fixed Articles date filter
### Testing Instructions
Create a new featured article.
Open the page in the front-end `index.php?option=com_content&view=featured&date_filtering=relative&relative_date=1`
or the page `index.php?option=com_content&view=featured&date_filtering=range&start_date_range=2017-10-12`
### Expected result
In the first case only articles created in the last 24 hours are displayed
In the second case, only articles created since 12/10/2017 are displayed
### Actual result
All articles are displayed
#### With patch applied:
In the first case only articles created in the last 24 hours are displayed
In the second case, only articles created since 12/10/2017 are displayed
### System information (as much as possible)
Joomla! 3.8.1
### Additional comments
Category List page and Category Blog are affected by the same bug
### Documentation Changes Required

eshiol committed:
fixes date filtering

ghost commented:
I have tested this item :red_circle: unsuccessfully on 6ade43a17333e71a3e44547281f2a7e2815781d9<br /><br />With and -out PR all featured Articles are shown if using `index.php?option=com_content&view=featured&date_filtering=relative&relative_date=1`<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18298">issues.joomla.org/tracker/joomla-cms/18298</a>.</sub>

eshiol commented:
It's very strange. If I use `index.php?option=com_content&view=featured&date_filtering=relative&relative_date=2` without PR I have
```
SELECT DISTINCT a.id, a.title, a.alias, a.introtext, a.fulltext, a.checked_out, a.checked_out_time, a.catid, a.created, a.created_by, a.created_by_alias, 
  CASE WHEN c.published = 2 
  AND a.state > 0 THEN 2 WHEN c.published != 1 THEN 0 ELSE a.state END as state,
  CASE WHEN a.modified = '0000-00-00 00:00:00' THEN a.created ELSE a.modified END as modified, a.modified_by, uam.name as modified_by_name,
  CASE WHEN a.publish_up = '0000-00-00 00:00:00' THEN a.created ELSE a.publish_up END as publish_up,a.publish_down, a.images, a.urls, a.attribs, a.metadata, a.metakey, a.metadesc, a.access, a.hits, a.xreference, a.featured, a.language,  LENGTH(a.fulltext) AS readmore, a.ordering,c.title AS category_title, c.path AS category_route, c.access AS category_access, c.alias AS category_alias,c.published, c.published AS parents_published, c.lft,
  CASE WHEN a.created_by_alias > ' ' THEN a.created_by_alias ELSE ua.name END AS author,ua.email AS author_email,parent.title as parent_title, parent.id as parent_id, parent.path as parent_route, parent.alias as parent_alias

  FROM k3mhz_content AS a

  LEFT JOIN k3mhz_categories AS c 
  ON c.id = a.catid

  LEFT JOIN k3mhz_users AS ua 
  ON ua.id = a.created_by

  LEFT JOIN k3mhz_users AS uam 
  ON uam.id = a.modified_by

  LEFT JOIN k3mhz_categories as parent 
  ON parent.id = c.parent_id

  WHERE a.featured = 1 
  AND a.access IN (1,1,5) 
  AND c.access IN (1,1,5) 
  AND c.published = 1 
  AND a.state = 1 
  AND (a.publish_up = '0000-00-00 00:00:00' OR a.publish_up <= '2017-10-10 08:00:42') 
  AND (a.publish_down = '0000-00-00 00:00:00' OR a.publish_down >= '2017-10-10 08:00:42') 

  ORDER BY c.lft,  
  CASE WHEN a.publish_up = '0000-00-00 00:00:00' THEN a.created ELSE a.publish_up END  DESC , a.created DESC  
  LIMIT 9
```
With PR I have
```
SELECT DISTINCT a.id, a.title, a.alias, a.introtext, a.fulltext, a.checked_out, a.checked_out_time, a.catid, a.created, a.created_by, a.created_by_alias, 
  CASE WHEN c.published = 2 
  AND a.state > 0 THEN 2 WHEN c.published != 1 THEN 0 ELSE a.state END as state,
  CASE WHEN a.modified = '0000-00-00 00:00:00' THEN a.created ELSE a.modified END as modified, a.modified_by, uam.name as modified_by_name,
  CASE WHEN a.publish_up = '0000-00-00 00:00:00' THEN a.created ELSE a.publish_up END as publish_up,a.publish_down, a.images, a.urls, a.attribs, a.metadata, a.metakey, a.metadesc, a.access, a.hits, a.xreference, a.featured, a.language,  LENGTH(a.fulltext) AS readmore, a.ordering,c.title AS category_title, c.path AS category_route, c.access AS category_access, c.alias AS category_alias,c.published, c.published AS parents_published, c.lft,
  CASE WHEN a.created_by_alias > ' ' THEN a.created_by_alias ELSE ua.name END AS author,ua.email AS author_email,parent.title as parent_title, parent.id as parent_id, parent.path as parent_route, parent.alias as parent_alias

  FROM k3mhz_content AS a

  LEFT JOIN k3mhz_categories AS c 
  ON c.id = a.catid

  LEFT JOIN k3mhz_users AS ua 
  ON ua.id = a.created_by

  LEFT JOIN k3mhz_users AS uam 
  ON uam.id = a.modified_by

  LEFT JOIN k3mhz_categories as parent 
  ON parent.id = c.parent_id

  WHERE a.featured = 1 
  AND a.access IN (1,1,5) 
  AND c.access IN (1,1,5) 
  AND c.published = 1 
  AND a.state = 1 
  AND (a.publish_up = '0000-00-00 00:00:00' OR a.publish_up <= '2017-10-10 08:00:42') 
  AND (a.publish_down = '0000-00-00 00:00:00' OR a.publish_down >= '2017-10-10 08:00:42') 
  AND a.created >= DATE_SUB('2017-10-10 08:00:42', INTERVAL 2 DAY)

  ORDER BY c.lft,  
  CASE WHEN a.publish_up = '0000-00-00 00:00:00' THEN a.created ELSE a.publish_up END  DESC , a.created DESC  
  LIMIT 9
```
The PR adds the where clause `a.created >= DATE_SUB('2017-10-10 08:00:42', INTERVAL 2 DAY)`

anuragteapot commented:
![2](https://user-images.githubusercontent.com/30978328/31488804-e1093d10-af0c-11e7-93e7-c8498dc5fc1c.png)
![1](https://user-images.githubusercontent.com/30978328/31488805-e2cf6c00-af0c-11e7-992c-f21a4b353108.png)

I have tested this item :white_check_mark: successfully on 6ade43a17333e71a3e44547281f2a7e2815781d9<br /><br />I have tested this. Without, patch I got that all articles are displayed. But, with patch applied Only articles created in the last 24 hours are displayed if using   <code>index.php?option=com_content&view=featured&date_filtering=relative&relative_date=1</code>.

But, for this  <code>index.php?option=com_content&view=featured&date_filtering=range&start_date_range=(url_ecoded yesterday sql format)</code>

it's not working. With and -out PR all featured Articles are shown . <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18298">issues.joomla.org/tracker/joomla-cms/18298</a>.</sub>

ghost commented:
@eshiol can you please inform where to test as i had test on Backend.

csthomas commented:
Should not this be designed as in https://github.com/joomla/joomla-cms/pull/18248 ?
using  `$this->getUserStateFromRequest()`?

alikon commented:
sorry for the silly question,
but how do you set that options from backend ?

csthomas commented:
What I understand now, this is designed for (manage from) front-end, not back-end.
Visitors can create URLs like `...&date_field=modified&....` and slow down whole website, administrator won't manage this directly from menu.


IIRC relative filters does not  work in postgreSQL.

A little similar is in filter_tag, visitor can set own tag to filter but administrator can predefine some tags in menu item.

alikon commented:
yes, but how a normal user can set these parameters if not from backend ?

csthomas commented on components/com_content/models/category.php:
```diff
@@ -214,6 +214,12 @@ protected function populateState($ordering = null, $direction = null)
+		$this->setState('filter.date_field', $app->input->getString('date_field'));
+		$this->setState('filter.start_date_range', $app->input->getString('start_date_range'));
+		$this->setState('filter.end_date_range', $app->input->getString('end_date_range'));
+		$this->setState('filter.relative_date', $app->input->getInt('relative_date'));
```
@alikon You have to be joomla advanced user:) Link `/index.php?option=com_content&view=category&id=1&date_field=modified&relative_date=1` and I can see what you have done in this category from yesterday:)

alikon replied:
from fronted ?
make little sense to me

csthomas replied:
Yes, for me too, 
This is not directly related to this PR but I would suggest to move this filters from URL request to menu params, take a look at featured view for `featured_categories` field in advanced fieldset.

ghost commented:
how to go on with this PR?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18298">issues.joomla.org/tracker/joomla-cms/18298</a>.</sub>

csthomas commented:
Despite it may work as author described, for me this is a wrong direction.

Crawlers, bots should not have ability to explore joomla website by sending requests like:
*   `index.php?option=com_content&view=featured&date_filtering=relative&relative_date=1`
*   `index.php?option=com_content&view=category&id=1&date_field=modified&relative_date=1`
*   `index.php?option=com_content&view=featured&date_filtering=range&start_date_range=...`

This functionality should be **only** available in menu items and modules (they can use `$model->setState()`)

At now (before this PR) as the filtering dates does not work IMO this functionality may be limited to only modules and menu items.

ghost commented:
This PR will be closed at 26th November 2017 if there is no other Decision in the Meantime.

rdeutz commented:
filters should be set from the backend, I don't think this is the way to go

ghost commented:
Status is set on "Needs Review".

ladyjer commented:
I have tested this item :white_check_mark: successfully on 6ade43a17333e71a3e44547281f2a7e2815781d9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18298">issues.joomla.org/tracker/joomla-cms/18298</a>.</sub>

brianteeman commented:
Closed a maintainer has already said this it not a good idea
